### PR TITLE
[tver] Add support for TVer

### PR DESCRIPTION
**Pull request by tsukumijima**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

---

[TVer](https://tver.jp/) is Japan's largest free distribution service for TV programs, operated mainly by five Japanese TV stations.
It distributes various programs such as Japanese anime, drama, news, and variety shows, and has more than 10 million active users.
This pull request was opened to make TVer videos available for download on youtube-dl.

#### Note

 - TVer is a service for Japan, so I set _GEO_COUNTRIES = ['JP'] to disguise the X-Fowarded-For header as a Japanese IP, but it may still not be available for download from outside Japan.
- The test from test / test_download.py doesn't seem to handle encrypted HLS videos correctly and says if pycryptodome is not installed, an error "ERROR: hlsnative has detected features it does not support, extraction will be delegated to ffmpeg" will occur.
Also, if pycryptodome is installed, an error "Value Error: Data must be padded to 16 byte boundary in CBC mode" will occur. I'm not sure why.
To avoid this, I've added a'skip'to _TESTS to skip the test (although most TVer videos aren't available after a week or so, so after a while the test itself doesn't make sense...)
- I an using Google Translate. The text may be unnatural.

### CODE REVIEW

File: youtube_dl/extractor/tver.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                'title': '港時間　#49　神奈川県／リビエラシーボニアマリーナ　9月18日(金)放送分',
+                'description': '【毎週金曜 よる12時15分から放送】\n\n日本のヨット文化 を育んできた三浦半島の西海岸、小網代湾にあるリビエラシーボニアマリーナ。昨年から始まったSailGPの日本チームを率いるヨット界のレジェンドに会いました。',
+                'thumbnail': 'https://cf-images.ap-northeast-1.prod.boltdns.net/v1/jit/4394098883001/904361ca-40d3-4028-8478-8916b9a0ff49/main/1920x1080/58s80ms/match/image.jpg',
+                'duration': 116.16,
+                'timestamp': 1600052421,
+                'upload_date': '20200914',
+                'uploader_id': '4394098883001',
+            },
+            'skip': 'Running from test_download.py doesn\'t seem to be able to handle encrypted HLS videos',
+        },
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:62]: Remove duplicate tests.

File: youtube_dl/extractor/tver.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import re
+
+from .brightcove import BrightcoveNewIE
+
+
+class TVerIE(BrightcoveNewIE):
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:9]: Do not inherit brightcove, you must delegate instead.

File: youtube_dl/extractor/tver.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        {
+            'url': 'https://tver.jp/feature/f0057485',  # 'feature'
+            'md5': '1c1c09662252571992dee0441028b4ec',  # MD5 hash of a short video downloaded by running youtube-dl with the --test option
+            'info_dict': {
+                'id': 'f0057485',  # TVer ID
+                'display_id': 'ref:hanzawa_naoki---s2----323-001',  # Brightcove ID
+                'ext': 'mp4',
+                'title': '半沢直樹(新シリーズ)　第1話 子会社VS銀行!飛ばされた半沢の新たな下剋上が始まる',
+                'description': '大和田(香川照之)の不正を糾弾し､子会社へ出向を命じられた半沢直樹(堺雅人)は､東京セントラル証券営業企画部長に｡ある日1500億円超の買収案件が舞い込むが…｡',
+                'thumbnail': 'https://cf-images.ap-northeast-1.prod.boltdns.net/v1/jit/4031511847001/37b5f176-3989-48d9-81d1-4688e80c5531/main/1920x1080/34m10s16ms/match/image.jpg',
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:21]: No exact URLs.

File: youtube_dl/extractor/tver.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    _TESTS = [
+        {
+            'url': 'https://tver.jp/feature/f0057485',  # 'feature'
+            'md5': '1c1c09662252571992dee0441028b4ec',  # MD5 hash of a short video downloaded by running youtube-dl with the --test option
+            'info_dict': {
+                'id': 'f0057485',  # TVer ID
+                'display_id': 'ref:hanzawa_naoki---s2----323-001',  # Brightcove ID
+                'ext': 'mp4',
+                'title': '半沢直樹(新シリーズ)　第1話 子会社VS銀行!飛ばされた半沢の新たな下剋上が始まる',
+                'description': '大和田(香川照之)の不正を糾弾し､子会社へ出向を命じられた半沢直樹(堺雅人)は､東京セントラル証券営業企画部長に｡ある日1500億円超の買収案件が舞い込むが…｡',
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:20]: md5

File: youtube_dl/extractor/tver.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    IE_DESC = 'TVer'
+
+    _VALID_URL = r'https?://(?:www\.)?tver\.jp/(corner|episode|feature)/(?P<id>f?[0-9]+)'
+    _GEO_COUNTRIES = ['JP']  # TVer service is limited to Japan only
+
+    BRIGHTCOVE_URL_TEMPLATE = 'http://players.brightcove.net/%s/default_default/index.html?videoId=ref:%s'
+
+    def _real_extract(self, url):
+
+        # extract video id
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:75]: Remove all useless comments.

File: youtube_dl/extractor/tver.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+        # brightcove url
+        brightcove_url = self.BRIGHTCOVE_URL_TEMPLATE % (brightcove_account_id, brightcove_video_id)
+
+        # debug output
+        if self._downloader.params.get('verbose', False):
+            self.to_screen('Video Information: %s' % video_info)
+            self.to_screen('Brightcove Account ID: %s' % brightcove_account_id)
+            self.to_screen('Brightcove Video ID: %s' % brightcove_video_id)
+            self.to_screen('Brightcove URL: %s' % brightcove_url)
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:100]: Remove all debug garbage.

File: youtube_dl/extractor/tver.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+        # evacuate _VALID_URL
+        _VALID_URL = self._VALID_URL
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:103]: Noway.

File: youtube_dl/extractor/tver.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+        # temporarily replace _VALID_URL
+        # prevent _VALID_URL from being the URL of Tver when executing the parent class's _real_extract () method
+        self._VALID_URL = r'https?://players\.brightcove\.net/(?P<account_id>\d+)/(?P<player_id>[^/]+)_(?P<embed>[^/]+)/index\.html\?.*(?P<content_type>video|playlist)Id=(?P<video_id>\d+|ref:[^&]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:107]: Noway.

File: youtube_dl/extractor/tver.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+        # get video information
+        info_dict = super(TVerIE, self)._real_extract(brightcove_url)
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:110]: Noway.

File: youtube_dl/extractor/tver.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        self._VALID_URL = _VALID_URL
+
+        # TVer ID
+        info_dict['id'] = video_id
+        # Brightcove ID
+        info_dict['display_id'] = brightcove_video_id
+        # select large thumbnail
+        info_dict['thumbnail'] = info_dict.get('thumbnail').replace('160x90', '1920x1080')
+        # desctiption
+        info_dict['description'] = description
```
Comment (dstftw) [inline youtube_dl/extractor/tver.py:127]: Noway. See other extractors on how to delegate properly.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/tver.py (hunk 1 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: It has been deleted.

File: youtube_dl/extractor/tver.py (hunk 3 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: Fixed.

File: youtube_dl/extractor/tver.py (hunk 2 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: TVer makes use of Brightcove, but some entities, such as descriptions, had to be extracted from TVer's site.
I didn't know how to overwrite the entity retrieved by BrightcoveNewIE in such a case, so I ended up with such a dirty implementation.
  In the return value, I found that it was possible to delegate in the correct way by setting '_type' to 'url_transparent' and setting the entity retrieved by TVerIE such as 'description', so I rewrote it significantly.
I didn't do enough preliminary research ... I'm sorry.

File: youtube_dl/extractor/tver.py (hunk 4 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: Fixed.

File: youtube_dl/extractor/tver.py (hunk 5 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: Obviously excessive comments have been removed.
However, I think that the code will be difficult to understand if there are too few comments, so I left the comments that I felt necessary.

File: youtube_dl/extractor/tver.py (hunk 6 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: It has been deleted.

File: youtube_dl/extractor/tver.py (hunk 7 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: Fixed.

File: youtube_dl/extractor/tver.py (hunk 8 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: Fixed.

File: youtube_dl/extractor/tver.py (hunk 9 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: Fixed.

File: youtube_dl/extractor/tver.py (hunk 10 above)
Comment (tsukumijima) [reply youtube_dl/extractor/tver.py]: I think I have fixed all the points you pointed out. 
Thank you for your review.

**Comment by tsukumijima** [comment]
I'm sorry to be busy, Would you please review it again?
Or should I reopen the PR?

**Comment by Paun** [comment]
Oh finally!!

I am very glad to see this will be supported soon.

Nice work.

**Comment by tsukumijima** [comment]
The merged changes are a bit different from the code I wrote, but in any case, I'm happy to see TVer support added to youtube-dl.
I haven't confirmed if it works yet, but as far as I can see the code, it should work.

**Comment by tsukumijima** [comment]
I've read the code and it's certainly more verbose and nice than my code.
I also admire the ability to write this code in just a few hours after receiving your request.

**Pull request closed without merging**